mantidproject/mantid
# Engg: implement first GUI

Fixes #12610, and closes #9213 (initial ENGIN-X umbrella ticket). This introduces a first version of the Engg/ENGIN-X GUI. You can open it from the main menu bar: Interfaces -> Diffraction -> Engineering Diffraction

**To test**:
- Code review. this comes with a few tests on the GUI presenter which should be good enough to cover basic functionality and most error cases.
- A relatively simple manual test could be done following the steps given below.

Steps to use the interface:
- You need to set some parameters in the tab "Settings":
  - "Input directories", "Raw files": set the folder '\olympic\Babylon5\Public\fedemp\ENGIN-X\GUI_calib_folder` (or a copy of this folder with the same ENGINX run files included in it).
  - For a first time setup: you need to generate the 'pixel calibration' that is done every several years (full calibration). You can use the file included with the test data: '\olympic\Babylon5\Public\fedemp\ENGIN-X\GUI_calib_folder\ENGINX_full_pixel_calibration.csv`. In the "Settings" tab, "Pixel calibration", click browse and select that file.
  - Leave other options as defaults:
    - The template GSAS file should be automatically found in the sources (or where the EnggUtils script is found).
- Then you can use the Calibrate tab to make a calibration:
  - You can put anything in "RB Number", or leave it empty, as it is not effectively used at the moment.
  - Make new calibration box: enter Vanadium #: 236516, and Ceria #: 241391 (these values are provided by default on startup at the moment).
  - Click Calibrate.
  - The calibration should be generated and it should become the current calibration.
  - You can also load it using the button "Load existing calibration" in "Current Calibration"

Please consider this as a starting point. There are several points to improve in follow-up (and smaller) pull requests. In particular these need short-term improvement for sure:
- The documentation would need several clarifications and additions, but the current version at least describes the main actions and parameters.
- Vanadium and Ceria run numbers are entered as run numbers as in the initial plan - but I think this could work better with the MWRunFiles widget. We have to check this with Saurabh. Related to this, in this current version if you enter a wrong Ceria or Vanadium number and the file is not found you will see the error at the time you try to generate the new calibration, and not as soon as you enter the wrong (unavailable) number which I think is preferable. This needs improvement (or if we finally use the MWRunFiles widget this issue would be fixed for free).
- It follows an MVP pattern but without the M for model, as for now a handful of workspaces play the role of the model. It's probably a good idea to create a model class as it grows.
- The settings (settings tab) are persistent between Mantid sessions. In addition, some of the parameters of the calibration tab (example: current vanadium and ceria run numbers) should be made persistent as well, unless it could create confusion. We have to clarify this with scientists.
- A few more tests should be added (there are some comments about this in the code).
